Arrange for publishing the project as a container image #41

Antoineio opened the pull request:
Hello @gridhead 👋,
I am opening this PR to address issue #41.

This PR introduces two new files and modifies one existing file, as detailed below:

## `Dockerfile`

In the Dockerfile, i added a multi-stage Dockerfile using a builder image and a runtime image to reduce the final image size.

- **Golang Installation (Manual Mode):**  
Since the official `dnf` repository only provides Golang version 1.23.x :man_shrugging:, i opted to manually install Golang version 1.24.1. This ensures we can use the required version for the project.

- **Binary Renaming:**  
  I renamed the compiled binary to `metasource-cli` to avoid accidental copying issues. 
  An incorrect name could lead to errors like:
  ```
  metasource  | Starting service in dispense mode...
  metasource  | /entrypoint.sh: line 15: ./metasource: Is a directory
  metasource exited with code 126
  ```
  
## `entrypoint.sh`

I added the `entrypoint.sh` script to handle cases where the `dispense` command is run without a provisioned database. Without a database, the server responds correctly but returns 400 errors.
```
[antoine@fedora metasource (main)]$ go build -o testbin
[antoine@fedora metasource (main)]$ mkdir tmp
[antoine@fedora metasource (main)]$ ls -l tmp/
total 0
[antoine@fedora metasource (main)]$ ./testbin -location tmp/ dispense
"GET http://127.0.0.1:8080/ HTTP/1.1" from 127.0.0.1:57766 - 200 6775B in 91.222µs
"GET http://127.0.0.1:8080/rawhide/pkg/kernel-devel HTTP/1.1" from 127.0.0.1:57766 - 400 17B in 51.59µs
"GET http://127.0.0.1:8080/rawhide/changelog/systemd-networkd HTTP/1.1" from 127.0.0.1:57766 - 400 17B in 30.28µs
```

- **Temporary Download and Copy:**  
  The script downloads the database into a temporary folder first and then copies it to the designated directory. This two-step process avoids “device or resource busy.” errors
  Example, when i try to copy on the mounted folder :
  ```
  [antoine@fedora metasource (main)]$ docker container run --volume="/tmp/metasource_db:/metasource_db" --publish="8080:8080" --name="metasource" metasource:latest
  Database not found, downloading...
  Apr 10 20:58:54.115 INF 37 repository location(s) acquired
  Apr 10 20:58:54.115 ERR unlinkat //metasource_db: device or resource busy
  ```
- **Conditional Execution:**  
  If the user already provides a database volume, the script skips the download and directly executes the `dispense` command.
  _I made this script temporarily but I think this behavior should maybe be handled by code :thinking:_

## `README.md`

Finally, i updated the `README.md` to include, step-by-step instructions on running the container with Docker.


I hope this meets your needs ! Let me know if any adjustments are required 🙏

gridhead reviewed (changes requested):
Please reword your commit message https://github.com/gridhead/metasource/pull/47/commits/f2ccafd3eaa195167e31f6d0ac1682968babf681 to "Add `Dockerfile` and `entrypoint.sh` & update `README.md`". The general thumb rule to follow while writing commit messages is to describe what changes a commit makes to the repository rather than describing what you did to the commit in the simple present tense.

More resources here https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/

sdglitched commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+
+COPY --from=builder /app/metasource-cli /app/metasource-cli
+
+COPY entrypoint.sh /entrypoint.sh
+RUN chmod +x /entrypoint.sh
+
+EXPOSE 8080
+
+ENTRYPOINT ["/entrypoint.sh"]
```
@Antoineio, the size of the database is quite large (approx 16 GiB). I believe having a volume mount for this would had been a good choice. This would reduce the container size to only the functional code.

We can have entrypoint.sh as a standalone script which can used to fetch the database.

@gridhead, what are your thoughts on this?

gridhead replied:
Yeah, we do not want to rebuild the container and deploy the service as and when there is a new repository update made from the cronjob because that would result in some downtime to account for the duration of downloading and processing. Please use volume mounting instead like they suggested as we expect to run this cronjob once per hour in the production and staging environments.

Antoineio commented:
:wave:

@gridhead
As mentioned, i have reworded my commit 🙂.

@sdglitched, thanks a lot for your feedback ! i really appreciate It :pray: 
I'm waiting to hear @gridhead’s thoughts on your comments before making changes

gridhead commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+    dnf -y install wget createrepo_c-devel gcc tar gzip && \
+    dnf clean all
+
+RUN wget https://go.dev/dl/go1.24.1.linux-amd64.tar.gz && \
+    tar -C /usr/local -xzf go1.24.1.linux-amd64.tar.gz && \
+    rm -f go1.24.1.linux-amd64.tar.gz
```
Yeah, I am building on Rawhide so I have the most recent version of Go at my end but that is not a requirement here. Please feel free to use whatever recent version of Go is available in whatever recent version of Fedora Linux base image available. Installing using the source tarballs like this is strongly discouraged when a native package for Go is available.

gridhead commented on README.md:
````diff
@@ -91,3 +91,21 @@ https://metasource.gridhead.net/
     ```
 12. Consider contributing to the project with methods that you see feasible.
 
+### Using Docker
````
```suggestion
### Containerized deployment
```

gridhead commented on README.md:
```diff
@@ -91,3 +91,21 @@ https://metasource.gridhead.net/
 
+### Using Docker
+
+1.  Clone the repository contents to your local projects directory.
```
Please start with providing instructions to install `podman`.

```
$ sudo dnf install podman
```

Then you can proceed with the instructions that you have mentioned.

gridhead commented on README.md:
````diff
@@ -91,3 +91,21 @@ https://metasource.gridhead.net/
+    ```
+3.  Build the image.
+    ```
+    $ docker image build . -t metasource:latest
````
I gotta confirm if this specification is 1:1 compatible with the `podman` API.

```suggestion
    $ podman image build . -t metasource:latest
```

gridhead commented on README.md:
````diff
@@ -91,3 +91,21 @@ https://metasource.gridhead.net/
+    ```
+4.  Run the container.
+    ```
+    $ docker container run --volume="/tmp/metasource_db:/metasource_db" --publish="8080:8080" --name="metasource" --detach metasource:latest
````
Make this span across multiple lines for enhanced readability.

```suggestion
    $ podman run \
          --volume="/tmp/metasource_db:/metasource_db" \
          --publish="8080:8080" \
          --name="metasource" \
          --detach metasource:latest
```

gridhead commented on entrypoint.sh:
```diff
@@ -0,0 +1,18 @@
+
+# If database is empty download it
+
+if [ -z "$(ls -A /metasource_db)" ]; then
+    echo "Database not found, downloading..."
+    mkdir -p /tmp/metasource_db
+    ./metasource-cli -location /tmp/metasource_db database
+    cp -r /tmp/metasource_db/* /metasource_db/
+    rm -rf /tmp/metasource_db
+fi
```
Instead of this, may I suggest that you leave the entrypoint of the container without mentioning a specific subcommand and that way, people can bend it to fill the requirements of both the `database` and `dispense` subcommands?

gridhead reviewed (changes requested):
@Antoineio, thanks for the changes so far - here are some more.

Antoineio commented:
Hello @gridhead & @sdglitched :wave: 

Thanks again for all your feedback i tried to work on it. I included all the changes in my last [commit](https://github.com/gridhead/metasource/pull/47/commits/2551d0cc7d760f890b1ba484f9c9e6596c0ded2e)

## Changes 
As the `go.mod` file was previously updated _(thanks to @sdglitched :muscle: )_, i simply removed the manual Go installation step from the Dockerfile, opting to install Go through the package manager (`dnf`).

I then added directives to creating the directory `/app/metasource_db` within the Dockerfile. This directory is used as a mount point for a **persistent volume**.
Additionally, i took this opportunity to rename the binary from `metasource-cli` to `meta` to align with changes introduced in previous commits.

Finally, i removed all instructions related to entrypoints scripts, and the `ENTRYPOINT` as we had discussed

### Issues encountered during testing

During my tests, i encountered several errors related to Podman volume mounting.

The image was built successfully:

```
$ podman image build . -t metasource:latest
[...]
Successfully tagged localhost/metasource:latest
```

However, when attempting to run the container with a local directory mounted as a volume, i received a **permissions denied** error :sweat: :
```
$ podman container run -v /home/antoine/tmp/metasource/:/app/metasource_db --name metasource metasource:latest ./meta -location /app/metasource_db database
Apr 15 22:01:53.067 INF 41 repository location(s) acquired
Apr 15 22:01:53.067 ERR openfdat /app/metasource_db: permission denied
```
Then, when using a named Podman volume, i encountered a different error :sob: :
```
$ podman container run -v metasource:/app/metasource_db --name metasource metasource:latest ./meta -location /app/metasource_db database
Apr 15 22:03:42.764 INF 41 repository location(s) acquired
Apr 15 22:03:42.764 ERR unlinkat /app/metasource_db: device or resource busy
```
Despite this error, the Podman volume was successfully created :thinking: :
```
$ podman volume ls
DRIVER      VOLUME NAME
local       metasource
```

### Potential cause

Upon researching online, i discovered that the error `unlinkat: device or resource busy` occurred because the code attempted to delete `/app/metasource_db` directory, which was a mounted volume.

I think, here was the problematic initial code:

```
if _, expt = os.Stat(loca); !os.IsNotExist(expt) {
    expt = os.RemoveAll(loca)
    if expt != nil {
        return expt
    }
    return InitPath(loca)
}
```

### Try to implemented fix

To address this issue, i try to update the code to delete only the contents of the existing directory without affecting the mount point itself :
```

func InitPath(loca string) error {
	var expt error

	_, expt = os.Stat(loca)
	if os.IsNotExist(expt) {
		expt = os.MkdirAll(loca, 0755)
		if expt != nil {
			return expt
		}
	} else {
		// If the folder already exists, delete its contents without affecting the mount point
		entries, err := os.ReadDir(loca)
		if err != nil {
			return err
		}
		for _, entry := range entries {
			path := fmt.Sprintf("%s/%s", loca, entry.Name())
			err = os.RemoveAll(path)
			if err != nil {
				return err
			}
		}
	}

	// Create required subdirectories
	expt = os.MkdirAll(fmt.Sprintf("%s/sxml", loca), 0755)
	if expt != nil {
		return expt
	}
	expt = os.MkdirAll(fmt.Sprintf("%s/comp", loca), 0755)
	if expt != nil {
		return expt
	}

	slog.Log(nil, slog.LevelWarn, "Directories initialized")
	return nil
}
```

This change ensures the directory content is cleaned without encountering errors related to mounted volumes.

After rebuilding the image and rerunning tests with a named Podman volume, i encountered no more issues:

```
$ podman container run -v metasource:/app/metasource_db --name metasource metasource:latest ./meta -location /app/metasource_db database
Apr 15 22:15:24.951 INF 41 repository location(s) acquired
Apr 15 22:15:24.952 WRN Directories initialized
[...]
Apr 15 22:19:18.701 INF [koji] Database indexing complete
Apr 15 22:19:20.431 WRN Directories removed
```

Additionally, i successfully ran the web interface and updated the database at the same time without issues.
```
[antoine@fedora ~ ]$ podman container run --rm -v metasource:/app/metasource_db --name metasource-updater metasource:latest ./meta -location /app/metasource_db database
Apr 15 22:29:54.046 INF 41 repository location(s) acquired
Apr 15 22:29:54.047 WRN Directories initialized
[antoine@fedora metasource (main)]$ podman container ls
CONTAINER ID  IMAGE                        COMMAND               CREATED         STATUS         PORTS                   NAMES
c02facf2bb6c  localhost/metasource:latest  ./meta -location ...  7 minutes ago   Up 7 minutes   0.0.0.0:8080->8080/tcp  metasource
9f5b7d67f779  localhost/metasource:latest  ./meta -location ...  37 seconds ago  Up 37 seconds  8080/tcp                metasource-updater
```

Although my approach _(directly modifying the code to solve the problem)_ works, i suspect it's probably not the best one :worried: . Indeed, i'm still a beginner, so i may have missed something, and i suspect improvements are needed.

I look forward to your feedback. :pray:

sdglitched reviewed (changes requested):
Hey @Antoineio,

I'll share some pointers with you related to contributing to Open Source:
1. Always branch out from `main` while adding a feature or fixing an issue
2. Have concise commit title under 50 characters and if additional information is needed, add it in commit description 
3. Having multiple commits is ok but they should make sense
In your case you could have had two commits, one for creating the docker file and another for changing the README
4. Do not include any unwanted commits (like the merge commit you added)

Please make these changes.

I've added a new PR #54 which might resolve the mounting issue you had. Once it is merged in main by @gridhead you can rebase the changes and then try your changes. But dont wait for it get merged and make the changes in mentioned above.

gridhead reviewed (changes requested):
@Antoineio, please rebase your current branch atop the updated `main` branch and I will give it another look.

sdglitched commented:
@Antoineio, after rebasing the from the main, please rollback the changes you did in the codebase for mounting issue and give it it a shot.
The containerisation should work without the issue you were facing.

Antoineio commented:
Hello @sdglitched and @gridhead :wave:  
Thank you for your feedback. I have tried to implement your recommendations in my last commits.
I hope this meets your expectations :pray:
